src/main/scala/com/fulcrumgenomics/rnaseq/AnnotateByIntervalList.scala

```diff
@@ -0,0 +1,246 @@
+    val templateIntervals = records.filter(record => {
+      if (record.mateMapped) {
+        val equalStartTieBreaker =
+          if (Math.min(record.start, record.end) == Math.min(record.mateStart, record.mateEnd.getOrElse(Int.MaxValue))) {
+            record.firstOfPair
+          } else {
+            true
+          }
+        // Take only the first read by coordinate order
+        Math.min(record.start, record.end) < Math.min(record.mateStart, record.mateEnd.getOrElse(Int.MaxValue)) &&
+        // if mate start == start pick first of pair
+        equalStartTieBreaker
+      } else {
+        //If we are single ended then just take the record
+        true
+      }
+    }).map(record =>
+      // Interval should be end of read one to the beginning of read two
+      new Interval(
+        record.refName,
+        Math.max(record.start, record.end) + 1,
+        Math.min(record.mateStart, record.mateEnd.getOrElse(Int.MaxValue)) - 1
+      )
+    )
+
+    var numLociMultiMatched: Long = 0
+
+    templateIntervals.foreach { templateInterval =>
+      // Overlaps should only be flagged when they enclose the template bases.
+      val overlaps = overlapDetector.getOverlaps(templateInterval).toSeq
+        .filter(interval => CoordMath.encloses(interval.getStart, interval.getEnd,
+          templateInterval.getStart, templateInterval.getEnd))
+
+      if (overlaps.isEmpty) {
+        baseCounter.count(this.unmatched, templateInterval.getLengthOnReference)
+        templateCounter.count(this.unmatched, 1)
+      } else {
+        val names: Seq[String] = overlaps.map(_.getName).distinct
+
+        def countForNames(namesToCount: Seq[String]): Unit = {
+          namesToCount.foreach { name =>
+            val bases = overlaps.filter(_.getName == name).map(_.getIntersectionLength(templateInterval)).sum
+            baseCounter.count(name, bases)
+            templateCounter.count(name, 1)
+            // If we only have one enclosing interval count unique bases and templates as well.
+            if (namesToCount.length == 1) {
+              uniqueBaseCounter.count(name, bases)
+              uniqueTemplateCounter.count(name, 1)
+            }
+          }
+        }
+
+        if (names.length > 1) {
+          // If there are multiple enclosing intervals by name use an associate strategy.
+          overlapAssociationStrategy match {
+            case Duplicate      => countForNames(names)
+            case MostSpecific   => countForNames(Seq(overlaps.minBy(_.length).getName))
+            case LeastSpecific  => countForNames(Seq(overlaps.maxBy(_.length).getName))
+          }
+          numLociMultiMatched += 1
+        } else {
+          // When there is only one interval matched we associate all intersected bases with the single interval
+          countForNames(names)
+        }
+      }
+      progress.record(templateInterval.getContig, templateInterval.getStart)
+    }
+    reader.close()
```

Comment:
You'll need to import `SafelyCloseable` from `FgbioDef` for the second line to work
```suggestion
    progress.logLast()
    reader.safelyClose()
```